actors/runtime/src/runtime/fvm.rs

```diff
@@ -263,28 +261,40 @@ where
         value: TokenAmount,
     ) -> Result<RawBytes, ActorError> {
         if self.in_transaction {
-            return Err(actor_error!(SysErrIllegalActor; "runtime.send() is not allowed"));
+            return Err(actor_error!(unspecified; "runtime.send() is not allowed"));
         }
-        // TODO: distinguish between "syscall" errors and actor exit codes.
         match fvm::send::send(&to, method, params, value) {
             Ok(ret) => {
                 if ret.exit_code.is_success() {
                     Ok(ret.return_data)
                 } else {
-                    Err(ActorError::from(ret.exit_code))
+                    // The returned code can't be simply propagated as it may be a system exit code.
+                    Err(ActorError::internal_send_aborted(format!(
+                        "send to {} method {} aborted with code {}",
+                        to, method, ret.exit_code
+                    )))
                 }
             }
             Err(err) => Err(match err {
+                // Some of these errors are from operations in the Runtime or SDK layer
+                // before or after the underlying VM send syscall.
```

Comment:
See my commentary here as I was trying to get to grips with things. I'm reaching the conclusion that error numbers are not good as error values, because context is lost and it's hard to know if the right message is being attached later.

Reply:
They seem to be only enough in the case we are okay with loosing that information, otherwise we must track additional information

Reply:
We're using "numbers" here (in the _syscall_ interface) because they're consensus critical and we want to keep things as "simple" as possible. If we can't distinguish between two cases, we should add additional exit codes for that. But I _think_ all cases are covered.

https://github.com/filecoin-project/fvm-specs/blob/main/07-errors.md#error-numbers-by-syscall should tell you _exactly_ what each error number means with respect to each syscall.

Reply:
> See my commentary here as I was trying to get to grips with things. I'm reaching the conclusion that error numbers are not good as error values, because context is lost and it's hard to know if the right message is being attached later.

Or are you referring to the exit code in `ActorError`? In that case, I agree. My attitude there is:

1. At the point where we can definitively say "we should exit with some specific exit code", we should just exit with that exit code (abort).
2. Before then, we should return descriptive (structured) errors.

Reply:
"Here" is _not_ the syscall interface, though. It's calling the SDK, which has more code and makes multiple syscalls. We need a descriptive structured error here. Propagating an error number is an antipattern.